`--local` flag for `ctl add`

kishansagathiya opened the pull request:
Fixes #848

hsanjuan commented on adder/single/dag_service.go:
```diff
@@ -62,19 +68,14 @@ func (dgs *DAGService) Add(ctx context.Context, node ipld.Node) error {
-	rootPin.Allocations = dgs.dests
+	if dgs.local {
+		return root, adder.Pin(ctx, dgs.rpcClient, rootPin)
+	}
```
The allocations should be respected. We are block-putting locally, but the file should be pinned normally as it would have been pinned if we were block-putting to other places.

kishansagathiya replied:
pin operation would take care of that since we care passing pin options with rootPin,
Checked it to be respecting options.

kishansagathiya replied:
https://github.com/ipfs/ipfs-cluster/pull/907/commits/ca81048ffa0b3cd62dc5e44c158f640f26264eec

hsanjuan commented on adder/single/dag_service.go:
```diff
@@ -30,22 +30,28 @@ type DAGService struct {
+	if dgs.local {
+		return adder.AddNodeLocal(ctx, dgs.rpcClient, node)
+	}
+
```
you need to do the blockallocate even if you don't use it to send the blocks to those destinations because you are running on local node.

hsanjuan replied:
Below, if you are in local-mode, your BlockAdder destinations are []("") (or better, modify the block adder to have the local parameter). There is no reason to add new code to add blocks to a place when you already have code to do that.

kishansagathiya replied:
can do this by passing an empty dest

hsanjuan commented on adder/util.go:
```diff
@@ -142,6 +142,21 @@ func Pin(ctx context.Context, rpc *rpc.Client, pin *api.Pin) error {
 }
 
+// AddNodeLocal puts the given node to the local peer.
+func AddNodeLocal(ctx context.Context, rpc *rpc.Client, node ipld.Node) error {
```
as noted above, this is duplicate code not really needed when you have a working blockAdder that does this.

hsanjuan commented on cluster.go:
```diff
@@ -1484,7 +1484,7 @@ func (c *Cluster) AddFile(reader *multipart.Reader, params *api.AddParams) (cid.
 		dags = sharding.New(c.rpcClient, params.PinOptions, nil)
 	} else {
-		dags = local.New(c.rpcClient, params.PinOptions)
+		dags = single.New(c.rpcClient, params.PinOptions, false)
```
You have the `params.Local` at hand to pass it in rather than hardcoding to false.

kishansagathiya commented:
Little trick I used to check if local add works
```
➜  ✗ ipfs-cluster-ctl add ~/Downloads/what.png 
added QmSq28uXPP3JMXTAZZSbRBaiUtqqJh8AYSg7d1NBgdPz4H what.png
➜  ✗ ipfs --api /ip4/127.0.0.1/tcp/5001 repo gc
removed QmPk85r5P6iaDwTC4VGwdvdNNuciQxBHA2tDM87JRwk3ie
➜  ✗ ipfs --api /ip4/127.0.0.1/tcp/5101 repo gc
removed QmPk85r5P6iaDwTC4VGwdvdNNuciQxBHA2tDM87JRwk3ie
➜  ✗ ipfs-cluster-ctl add --local ~/Downloads/what.png
added QmSq28uXPP3JMXTAZZSbRBaiUtqqJh8AYSg7d1NBgdPz4H what.png
➜  ✗ ipfs --api /ip4/127.0.0.1/tcp/5001 repo gc       
removed QmPk85r5P6iaDwTC4VGwdvdNNuciQxBHA2tDM87JRwk3ie
➜  ✗ ipfs --api /ip4/127.0.0.1/tcp/5101 repo gc       
➜  ✗ 
```

local ipfs would have extra blocks to gc

hsanjuan commented on adder/single/dag_service.go:
```diff
@@ -18,7 +18,7 @@ import (
 var errNotFound = errors.New("dagservice: block not found")
 
-var logger = logging.Logger("localdags")
+var logger = logging.Logger("singledags")
```
Need to update `logging.go` too.

kishansagathiya replied:
Made this change in the latest commit, but interestingly this log is never used.

hsanjuan commented on cmd/ipfs-cluster-ctl/main.go:
```diff
@@ -338,6 +338,10 @@ content.
 				},
+				cli.BoolFlag{
+					Name:  "local",
+					Usage: "Puts blocks only in local IPFS peer and then pin as per PinOptions",
```
It's not clear what `PinOptions` is here. Better `Add to local peer but pin normally`.

And then update the related paragraph in the long description:

```
Once the adding process is finished, the content is fully added to all
allocations and pinned in them. This makes cluster add slower than a local
ipfs add, but the result is a fully replicated CID on completion.
If you prefer faster adding, use the --local flag to add directly to the local IPFS node
and pin in the destinations after that. Note that the local IPFS node may not be
among the destinations, which will leave the unpinned content in it.
```
